haystack/utils/base_serialization.py

```diff
@@ -52,3 +52,84 @@ class does not have a `from_dict` method.
         raise DeserializationError(f"Class '{data['type']}' does not have a 'from_dict' method")
 
     return obj_class.from_dict(data["data"])
+
+
+def serialize_value(value: Any) -> Any:
+    """
+    Serializes a value into a format suitable for storage or transmission.
+
+    Handles various types including:
+    - Haystack dataclass objects (Answer, Document, etc.)
+    - Primitive types (returned as is)
+    - Lists of primitives (returned as is)
+    - Lists of complex types (recursively serialized)
+    - Dictionaries (recursively serialized)
+
+    :param value: The value to serialize
+    :returns: The serialized representation of the value
+    """
+
+    if hasattr(value, "to_dict") and callable(getattr(value, "to_dict")):
+        serialized_value = value.to_dict()
+        serialized_value["_type"] = value.__class__
+        return serialized_value
+
+    # this is a hack to serialize inputs that don't have a to_dict
+    elif hasattr(value, "__dict__"):
+        return {"_type": value.__class__, "attributes": value.__dict__}
+
+    # recursively serialize all inputs in a dict
+    elif isinstance(value, dict):
+        return {k: serialize_value(v) for k, v in value.items()}
+
+    # recursively serialize all inputs in lists or tuples
+    elif isinstance(value, list):
+        return [serialize_value(item) for item in value]
+
+    return value
+
+
+# pylint: disable=too-many-return-statements
+def deserialize_value(value: Any) -> Any:
+    """
+    Deserializes a value from its serialized representation.
+
+    Handles various types including:
+    - Haystack dataclass objects (Answer, Document, etc.)
+    - Primitive types (returned as is)
+    - Lists of primitives (returned as is)
+    - Lists of complex types (recursively deserialized)
+    - Dictionaries (recursively deserialized)
+
+    :param value: The serialized value to deserialize
+    :returns: The deserialized value
+    """
+
+    # None or primitive types are returned as is
+    if not value or isinstance(value, (str, int, float, bool)):
+        return value
+
+    # list of primitive types are returned as is
+    if isinstance(value, list) and all(isinstance(i, (str, int, float, bool)) for i in value):
+        return value
```

Comment:
probably could expand this slightly to include other container types like set and tuple
```suggestion
    # list of primitive types are returned as is
    if isinstance(value, (list, tuple, set)) and all(isinstance(i, (str, int, float, bool)) for i in value):
        return value
```

Reply:
Actually you can drop this block all together once you make the change in this [comment](https://github.com/deepset-ai/haystack/pull/9345/files#r2090982591) Since if we always recursively go into a list we will eventually hit a python primitive type and return back up. 

So no need to test for this edge case.